contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,316 @@
+        // VolumeSnapshotConditionReady is added when the snapshot has been successfully created and is ready to be used.
+        VolumeSnapshotConditionReady VolumeSnapshotConditionType = "Ready"
+        // VolumeSnapshotConditionError means an error occurred during snapshot creation.
+        VolumeSnapshotConditionError VolumeSnapshotConditionType = "Error"
+)
+
+// VolumeSnapshot Condition describes the state of a volume snapshot at a certain point.
+type VolumeSnapshotCondition struct {
+	// Type of replication controller condition.
+	Type VolumeSnapshotConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=VolumeSnapshotConditionType"`
+	// Status of the condition, one of True, False, Unknown.
+	Status core_v1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+	// The last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason" protobuf:"bytes,4,opt,name=reason"`
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string `json:"message" protobuf:"bytes,5,opt,name=message"`
+}
+
+```
+
+#### The `VolumeSnapshotData` Object
+
+```GO
+
+// +genclient=true
+// +nonNamespaced=true
+
+// VolumeSnapshotData represents the actual "on-disk" snapshot object
+type VolumeSnapshotData struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	Metadata metav1.ObjectMeta `json:"metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotDataSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// VolumeSnapshotDataList is a list of VolumeSnapshotData objects
+type VolumeSnapshotDataList struct {
+        metav1.TypeMeta `json:",inline"`
+        Metadata        metav1.ListMeta      `json:"metadata"`
+        Items           []VolumeSnapshotData `json:"items"`
+}
+
+// The desired state of the volume snapshot
+type VolumeSnapshotDataSpec struct {
+	// Source represents the location and type of the volume snapshot
+	VolumeSnapshotDataSource `json:",inline" protobuf:"bytes,1,opt,name=volumeSnapshotDataSource"`
+
+	// VolumeSnapshotRef is part of bi-directional binding between VolumeSnapshot
+	// and VolumeSnapshotData
+	// +optional
+	VolumeSnapshotRef *core_v1.ObjectReference `json:"volumeSnapshotRef" protobuf:"bytes,2,opt,name=volumeSnapshotRef"`
+
+	// PersistentVolumeRef represents the PersistentVolume that the snapshot has been
+	// taken from
+	// +optional
+	PersistentVolumeRef *core_v1.ObjectReference `json:"persistentVolumeRef" protobuf:"bytes,3,opt,name=persistentVolumeRef"`
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
```

Comment:
Since only CSI is going to be supported, this struct should probably contain only `CSISnapshotSource`:

```go
type VolumeSnapshotDataSource struct {
    CSI *CSISnapshotSource
}

type CSISnapshotSource struct {
    ID string
}
```

Reply:
Sure.